### PR TITLE
Allow static initialization for BPF_MAP_TYPE_HASH_OF_MAPS map with integer as key_size

### PR DESCRIPTION
**Pull request by shpalani**
## Description

1. In ebpf_native.c : Check map-in-map is Hash of Maps with integer as key_size. If true, allow initialization.
2. Added test cases in libbpf_tests.cpp.

## Testing

_Do any existing tests cover this change? No
Are new tests needed? Yes

Manual test:
```
PS C:\Temp> netsh ebpf add program .\hash_of_map_in_array_map.sys
Loaded with ID 196609

PS C:\Temp> dir .\hash_of_map_in_array_map.sys


    Directory: C:\Temp


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        10/5/2023   9:09 AM          11664 hash_of_map_in_array_map.sys


PS C:\Temp> netsh ebpf show programs

    ID  Pins  Links  Mode       Type           Name
======  ====  =====  =========  =============  ====================
196609     1      1  NATIVE     xdp            lookup

PS C:\Temp> netsh ebpf show maps

                              Key  Value      Max  Inner
     ID            Map Type  Size   Size  Entries     ID  Pins  Name
=======  ==================  ====  =====  =======  =====  ====  ========
 131073        hash_of_maps     2      4    65536  65537     0  outer_map

PS C:\Temp>
```
## Documentation

_Is there any documentation impact for this change? No

## Installation

_Is there any installer impact for this change? No

### CODE REVIEW

**Review by dthaler — DISMISSED** [review]

File: libs/execution_context/ebpf_native.c [hunk 1]
```diff
@@ -921,16 +921,22 @@ _ebpf_native_set_initial_map_values(_Inout_ ebpf_native_module_t* module)
             }
 
             ebpf_handle_t handle_to_insert = ebpf_handle_invalid;
-
-            if (native_map_to_update->entry->definition.type == BPF_MAP_TYPE_ARRAY_OF_MAPS) {
+            ebpf_map_definition_in_file_t map_def = native_map_to_update->entry->definition;
+            uint32_t key_size = map_def.key_size;
+            bool is_hash_of_maps_with_int_key = (map_def.type == BPF_MAP_TYPE_HASH_OF_MAPS) &&
+                                                (key_size == sizeof(uint8_t) ||
```
Comment (shankarseal) [inline libs/execution_context/ebpf_native.c:927]: what if key size is uint8_t and max_entries exceeds 256 etc.? Does this code handle that? @Alan-Jowett  can you confirm if Linux mandates key_size to be strictly 4-bytes in this case?

File: tests/sample/hash_of_map_in_array_map.c [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+//     generate_expected_bpf2c_output.ps1
+//
+// Usage:
+// .\scripts\generate_expected_bpf2c_output.ps1 <build_output_path>
+// Example:
+// .\scripts\generate_expected_bpf2c_output.ps1 .\x64\Debug\
+
+#include "bpf_helpers.h"
+
+#define LB_MAGLEV_MAP_MAX_ENTRIES 65536
```
Comment (shankarseal) [inline tests/sample/hash_of_map_in_array_map.c:16]: Please change the constant to have some other identifier than maglev.

File: tests/sample/hash_of_map_in_array_map.c [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+} outer_map SEC(".maps") = {
+    .values = {&inner_map},
+};
+
+SEC("xdp_prog") int lookup(struct xdp_md* ctx)
+{
+    uint16_t outer_key = 0;
+    void* maglev_inner_map = bpf_map_lookup_elem(&outer_map, &outer_key);
+    if (maglev_inner_map) {
+        uint32_t magler_inner_key = 0;
```
Comment (shankarseal) [inline tests/sample/hash_of_map_in_array_map.c:43]: typo? `magler` ?

File: tests/sample/hash_of_map_in_array_map.c (hunk 2 above)
Comment (saxena-anurag) [reply tests/sample/hash_of_map_in_array_map.c:16]: I think we should remove maglev name from the file.

Comment (shpalani) [reply tests/sample/hash_of_map_in_array_map.c:16]: Can you clarify the reason not to have 'maglev' string?

File: libs/execution_context/ebpf_native.c (hunk 1 above)
Comment (shpalani) [reply libs/execution_context/ebpf_native.c:927]: Can you please clarify the concern on the max_entries and key_size set?

File: tests/sample/hash_of_map_in_array_map.c (hunk 3 above)
Comment (shpalani) [reply tests/sample/hash_of_map_in_array_map.c:43]: Fixed.

File: libs/execution_context/ebpf_native.c (hunk 1 above)
Comment (shpalani) [reply libs/execution_context/ebpf_native.c:927]: @shankarseal : Could you please help clarify the discussed suggestion for the value_size?

For static initialization of hash-of-maps,
1. Add key_size == sizeof(int) or sizeof(long)  (so, it is only 4 bytes allowed).
2. value_size == ???

Comment (shpalani) [reply libs/execution_context/ebpf_native.c:927]: With the suggested changes of key_size == sizeof(int), adding value into the inner map array is failing. Hence the suggestion does not work.

Comment (shankarseal) [reply libs/execution_context/ebpf_native.c:927]: If the key size is x, there can be at most 2^8x entries in the map which must be checked.
Also, it is not enough to only check the length of the key; but the type should be an integer. Not sure if/how that can be verified here.

Comment (saxena-anurag) [reply libs/execution_context/ebpf_native.c:927]: If bpf2c has that info (from parsing btf data), I think we need to change bpf2c to add that information in native driver, and then this code can check it.

**Comment by Alan-Jowett** [comment]
@shpalani What is the status of this PR? Has it been abandoned? If so, please close. If not, please address any comments and mark as ready for review.

**Comment by shpalani** [comment]
Closing this PR, as it is committed through
https://github.com/microsoft/ebpf-for-windows/pull/3211

**Pull request closed without merging**